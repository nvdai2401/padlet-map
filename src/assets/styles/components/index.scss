@import './top-bar';
@import './folk-me';
@import './spinner';
@import './post-content';
@import './post-popup';
@import './mini-post';
@import './post-preview';
@import './post-expanded';
@import './map-control';
@import './map-view';
