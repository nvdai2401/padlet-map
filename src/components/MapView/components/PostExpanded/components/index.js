export { default as PostSlider } from './PostSlider';
