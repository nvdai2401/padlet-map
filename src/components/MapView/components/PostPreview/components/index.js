export { default as MiniPost } from './MiniPost';
