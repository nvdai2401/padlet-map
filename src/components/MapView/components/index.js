export { default as Post } from './Post';
export { default as PostPreview } from './PostPreview';
export { default as PostExpanded } from './PostExpanded';
