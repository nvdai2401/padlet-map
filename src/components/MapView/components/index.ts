export { default as PostContent } from './PostContent';
export { default as PostPopup } from './PostPopup';
export { default as PostPreview } from './PostPreview';
export { default as PostExpanded } from './PostExpanded';
export { default as MapControl } from './MapControl';
