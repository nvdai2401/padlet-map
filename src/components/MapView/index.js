export { default } from './MapView.vue';
