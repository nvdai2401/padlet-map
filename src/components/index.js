export { default as MapView } from './MapView';
export { default as TopBar } from './TopBar';
