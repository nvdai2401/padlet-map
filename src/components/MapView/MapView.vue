<template>
  <div class="map-view">
    <gmap-map
      ref="gmap"
      class="gmap-container"
      :zoom="initialZoom"
      :center="center"
      :options="mapOptions"
      @click="handleOnClickMap"
    >
      <gmap-marker
        ref="markers"
        v-for="marker in Object.values(markers)"
        :key="marker.id"
        :icon="{
          url: require(`src/assets/images/markers/${marker.color}_marker.webp`),
        }"
        :clickable="true"
        :position="{
          lat: marker.location_point.latitude,
          lng: marker.location_point.longitude,
        }"
        @click="handleOnClick(marker.id)"
        @mouseover="handleOnMouseOver(marker.id)"
        @mouseout="handleOnMouseOut(marker.id)"
      >
        <gmap-info-window
          :opened="marker.is_post_visible"
          @closeclick="is_post_visible = false"
        >
          <post
            v-if="marker.is_post_visible"
            :post-info="marker"
            :onExpandPost="handleOnExpandPost"
          />
        </gmap-info-window>
      </gmap-marker>
    </gmap-map>

    <div :class="['map-control', postPreviewVisible ? 'preview-shown' : '']">
      <button
        class="button map-center-control"
        @click="postPreviewVisible = !postPreviewVisible"
      >
        <font-awesome-icon :icon="['fas', 'bars']" />
      </button>

      <div class="group-button">
        <button class="button" @click="fitBounds">
          <font-awesome-icon :icon="['fas', 'expand']" />
        </button>
        <button class="button" @click="zoomIn">
          <font-awesome-icon :icon="['fas', 'plus']" />
        </button>
        <button class="button" @click="zoomOut">
          <font-awesome-icon :icon="['fas', 'minus']" />
        </button>
      </div>
    </div>
    <transition-group tag="div" name="slide">
      <post-preview
        v-show="postPreviewVisible && Object.keys(markers).length > 0"
        key="post-preview"
        :posts="markers"
        :onClick="handleOnClick"
        :onMouseOver="handleOnMouseOver"
        :onMouseOut="handleOnMouseOut"
      />
    </transition-group>

    <post-expanded
      v-if="postExpandedVisible"
      :open="postExpandedVisible"
      :postId="activePost"
      :posts="Object.keys(markers)"
      :onClose="handleOnClosePostExpanded"
    />
  </div>
</template>

<script>
import { Post, PostPreview, PostExpanded } from "./components";
import {
  defaultLatLngBounds,
  viewportLatLngBounds,
  options,
  DEFAULT_CENTER,
  MIN_ZOOM,
  MAX_ZOOM,
} from "./mapProperty";
import { getMarkers } from "src/api";

export default {
  name: "map-view",
  components: {
    post: Post,
    "post-preview": PostPreview,
    "post-expanded": PostExpanded,
  },
  data() {
    return {
      initialZoom: MIN_ZOOM,
      center: DEFAULT_CENTER,
      infoWindowVisible: false,
      postPreviewVisible: false,
      postExpandedVisible: false,
      markers: {},
      mapOptions: null,
      activePost: null,
      activeMarkerColor: "",
      focusedPost: "",
    };
  },
  watch: {
    activePost(newPostId, oldPostId) {
      const restriction = {
        ...this.mapOptions.restriction,
        latLngBounds: viewportLatLngBounds,
      };
      // Hide old post
      if (oldPostId) {
        this.markers[oldPostId].is_post_visible = false;
      }
      // Hide current post and update map bounds to fit the viewport when click its marker
      if (!newPostId) {
        this.mapOptions = { ...this.mapOptions, restriction };
        return;
      }
      // Show post and change map bounds to default when click a marker
      this.markers[newPostId].is_post_visible = true;
      restriction.latLngBounds = defaultLatLngBounds;
      this.mapOptions = { ...this.mapOptions, restriction };
    },
  },
  created() {
    this.mapOptions = options;
  },
  async mounted() {
    await this.fetchMakers();
    this.fitBounds();
  },
  methods: {
    async fetchMakers() {
      const data = await getMarkers();
      const markers = {};

      for (let item of data) {
        markers[item.id] = { ...item.attributes, is_post_visible: false };
      }
      this.markers = markers;
    },
    handleOnClick(postId) {
      if (this.activePost === postId) {
        this.activePost = null;
        return;
      }
      this.activePost = postId;
    },
    handleOnMouseOver(postId) {
      if (postId !== this.focusedPost) {
        const PLACEHOLDER_PREFIX = "ph_";
        this.focusedPost = postId;
        this.activeMarkerColor = this.markers[postId].color;
        this.markers[postId].color =
          PLACEHOLDER_PREFIX + this.markers[postId].color;
      }
    },
    handleOnMouseOut(postId) {
      this.markers[postId].color = this.activeMarkerColor;
      this.activeMarkerColor = "";
      this.focusedPost = "";
    },
    handleOnClickMap() {
      this.activePost = null;
    },
    fitBounds() {
      this.$refs.gmap.$mapPromise.then((map) => {
        var bounds = new google.maps.LatLngBounds();
        Object.values(this.markers)
          .reverse()
          .forEach((marker) => {
            bounds.extend(
              new google.maps.LatLng(
                marker.location_point.latitude,
                marker.location_point.longitude
              )
            );
          });
        map.fitBounds(bounds);
      });
    },
    handleOnExpandPost() {
      this.postExpandedVisible = true;
    },
    handleOnClosePostExpanded() {
      this.postExpandedVisible = false;
      this.activePost = null;
    },
    zoomIn() {
      const currentZoom = this.$refs.gmap.$mapObject.zoom;

      if (currentZoom < MAX_ZOOM) {
        this.mapOptions = { ...this.mapOptions, zoom: currentZoom + 1 };
      }
    },
    zoomOut() {
      const currentZoom = this.$refs.gmap.$mapObject.zoom;

      if (currentZoom > MIN_ZOOM) {
        this.mapOptions = { ...this.mapOptions, zoom: currentZoom - 1 };
      }
    },
  },
};
</script>

<style lang="scss" scoped>
.map-view {
  height: 90vh;
  position: relative;

  .gmap-container {
    width: 100vw;
    height: 100%;
  }

  .preview-shown {
    transform: translateY(-74px);
  }

  .map-control {
    position: absolute;
    z-index: 3000;
    bottom: 48px;
    left: 12px;
    display: flex;
    flex-direction: column;

    transition: transform 0.25s cubic-bezier(0.4, 0, 0.2, 1);

    svg {
      height: 14px;
    }

    .button {
      width: 30px;
      height: 30px;
      display: flex;
      align-items: center;
      justify-content: center;
      color: #666666;
      background-color: #ffffff;
      transition: all 0.15s ease-out;
      cursor: pointer;
      border: 0;
      outline: unset;

      &:hover {
        svg {
          color: #000000;
        }
      }
    }

    .map-center-control {
      border-radius: 6px;
    }

    .group-button {
      margin-top: 6px;
      border-radius: 6px;
      overflow: hidden;
      box-shadow: 0 1px 1px rgba(0, 0, 0, 0.12);
      transition: all 0.1s cubic-bezier(0.4, 0, 0.2, 1);
    }
  }

  .slide-leave-active,
  .slide-enter-active {
    transition: 0.3s;
  }
  .slide-enter {
    transform: translate(0, 100%);
  }
  .slide-leave-to {
    transform: translate(0, 100%);
  }

  @media screen and (min-width: 768px) {
    .preview-shown {
      transform: unset;
    }

    .map-control {
      flex-direction: row;
      bottom: 26px;
      right: 12px;
      left: unset;

      svg {
        height: 14px;
      }

      .group-button {
        display: flex;
        margin-top: 0;
        margin-left: 12px;
      }
    }

    .slide-enter {
      transform: translate(100%, 0);
    }
    .slide-leave-to {
      transform: translate(100%, 0);
    }
  }
}
</style>
