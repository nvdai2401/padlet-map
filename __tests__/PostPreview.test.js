/* eslint-disable no-undef */
import { posts } from './shares/mockData';
import { mountWithProps } from './shares/utils';
import { PostPreview } from '@/components/MapView/components';

const propsData = {
  posts: posts,
  onMouseOverPost: jest.fn(),
  onMouseOutPost: jest.fn(),
  onClickPost: jest.fn(),
};
describe('<PostPreview />', () => {
  it(`should render ${posts.length} posts`, () => {
    const wrapper = mountWithProps(PostPreview, propsData);

    expect(wrapper.find('.post-count').text()).toBe(`${posts.length} posts`);
    expect(wrapper.findAll('.post-list__container').length).toBe(posts.length);
  });

  it(`should call onMouseOverPost when mouse over`, () => {
    const wrapper = mountWithProps(PostPreview, propsData);

    wrapper.find('.mini-post').trigger('mouseover');
    expect(propsData.onMouseOverPost).toHaveBeenCalled();
  });

  it(`should call onMouseOutPost when mouse out`, () => {
    const wrapper = mountWithProps(PostPreview, propsData);

    wrapper.find('.mini-post').trigger('mouseout');
    expect(propsData.onMouseOutPost).toHaveBeenCalled();
  });

  it(`should call onClickPost when mouse out`, () => {
    const wrapper = mountWithProps(PostPreview, propsData);

    wrapper.find('.mini-post').trigger('click');
    expect(propsData.onClickPost).toHaveBeenCalled();
  });
});
