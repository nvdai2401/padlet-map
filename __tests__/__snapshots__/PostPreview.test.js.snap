// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PostPreview /> should match snapshot 1`] = `
<div
  class="post-preview"
>
  <div
    class="post-count"
  >
    13 posts
  </div>
   
  <div
    class="flex-row post-list"
  >
    <div
      class="flex-row post-list__container"
    >
      <div
        class="flex-row justify-space-between mini-post"
      >
        <div
          class="flex-column post-title"
        >
          <span
            class="title"
            style="color: red;"
          >
            United States
          </span>
           
          <span
            class="subtitle"
          >
            <div>
              Came here to study in 2006.
            </div>
          </span>
        </div>
         
        <div
          class="post-image"
        >
          <img
            alt="United States"
            height="56px"
            loading="lazy"
            src="https://ik.imagekit.io/fireman35/pics/usa_98nvVlwOEN.jpg?tr=w-112"
            width="56px"
          />
        </div>
      </div>
       
      <div
        class="divider"
      />
    </div>
    <div
      class="flex-row post-list__container"
    >
      <div
        class="flex-row justify-space-between mini-post"
      >
        <div
          class="flex-column post-title"
        >
          <span
            class="title"
            style="color: green;"
          >
            Canada
          </span>
           
          <span
            class="subtitle"
          >
            <div>
              Came here in 2018 to chill for a weekend.
            </div>
          </span>
        </div>
         
        <div
          class="post-image"
        >
          <img
            alt="Canada"
            height="56px"
            loading="lazy"
            src="https://ik.imagekit.io/fireman35/pics/canada_GaDi1RWytiLm.jpg?tr=w-112"
            width="56px"
          />
        </div>
      </div>
       
      <div
        class="divider"
      />
    </div>
    <div
      class="flex-row post-list__container"
    >
      <div
        class="flex-row justify-space-between mini-post"
      >
        <div
          class="flex-column post-title"
        >
          <span
            class="title"
            style="color: blue;"
          >
            Mongolia
          </span>
           
          <span
            class="subtitle"
          >
            <div>
              Came here on vacation in 2009 (p.s. it's quite cold. Go in summer.)
            </div>
          </span>
        </div>
         
        <div
          class="post-image"
        >
          <img
            alt="Mongolia"
            height="56px"
            loading="lazy"
            src="https://ik.imagekit.io/fireman35/pics/mongolia_mWRhofTG4mL.jpg?tr=w-112"
            width="56px"
          />
        </div>
      </div>
       
      <div
        class="divider"
      />
    </div>
    <div
      class="flex-row post-list__container"
    >
      <div
        class="flex-row justify-space-between mini-post"
      >
        <div
          class="flex-column post-title"
        >
          <span
            class="title"
            style="color: green;"
          >
            Chile
          </span>
           
          <span
            class="subtitle"
          >
            <div>
              Came here to start a company in 2012.
            </div>
          </span>
        </div>
         
        <div
          class="post-image"
        >
          <img
            alt="Chile"
            height="56px"
            loading="lazy"
            src="https://ik.imagekit.io/fireman35/pics/chile_0qO775HZwWa.jpg?tr=w-112"
            width="56px"
          />
        </div>
      </div>
       
      <div
        class="divider"
      />
    </div>
    <div
      class="flex-row post-list__container"
    >
      <div
        class="flex-row justify-space-between mini-post"
      >
        <div
          class="flex-column post-title"
        >
          <span
            class="title"
            style="color: green;"
          >
            Brazil
          </span>
           
          <span
            class="subtitle"
          >
            <div>
              Came here for a conference in 2012.
            </div>
          </span>
        </div>
         
        <div
          class="post-image"
        >
          <img
            alt="Brazil"
            height="56px"
            loading="lazy"
            src="https://ik.imagekit.io/fireman35/pics/brazil_96W3F3LPa.jpg?tr=w-112"
            width="56px"
          />
        </div>
      </div>
       
      <div
        class="divider"
      />
    </div>
    <div
      class="flex-row post-list__container"
    >
      <div
        class="flex-row justify-space-between mini-post"
      >
        <div
          class="flex-column post-title"
        >
          <span
            class="title"
            style="color: red;"
          >
            France
          </span>
           
          <span
            class="subtitle"
          >
            <div>
              Came here to study in 2005.
            </div>
          </span>
        </div>
         
        <div
          class="post-image"
        >
          <img
            alt="France"
            height="56px"
            loading="lazy"
            src="https://ik.imagekit.io/fireman35/pics/france_9sri-tCHe.jpg?tr=w-112"
            width="56px"
          />
        </div>
      </div>
       
      <div
        class="divider"
      />
    </div>
    <div
      class="flex-row post-list__container"
    >
      <div
        class="flex-row justify-space-between mini-post"
      >
        <div
          class="flex-column post-title"
        >
          <span
            class="title"
            style="color: red;"
          >
            Singapore
          </span>
           
          <span
            class="subtitle"
          >
            <div>
              Came here to study in 2001.
            </div>
          </span>
        </div>
         
        <div
          class="post-image"
        >
          <img
            alt="Singapore"
            height="56px"
            loading="lazy"
            src="https://ik.imagekit.io/fireman35/pics/singapore_wrnhKpF1vKYY.jpg?tr=w-112"
            width="56px"
          />
        </div>
      </div>
       
      <div
        class="divider"
      />
    </div>
    <div
      class="flex-row post-list__container"
    >
      <div
        class="flex-row justify-space-between mini-post"
      >
        <div
          class="flex-column post-title"
        >
          <span
            class="title"
            style="color: pink;"
          >
            India
          </span>
           
          <span
            class="subtitle"
          >
            <div>
              Born here.
            </div>
          </span>
        </div>
         
        <div
          class="post-image"
        >
          <img
            alt="India"
            height="56px"
            loading="lazy"
            src="https://ik.imagekit.io/fireman35/pics/india_A4d15u4bMQ7.jpg?tr=w-112"
            width="56px"
          />
        </div>
      </div>
       
      <div
        class="divider"
      />
    </div>
    <div
      class="flex-row post-list__container"
    >
      <div
        class="flex-row justify-space-between mini-post"
      >
        <div
          class="flex-column post-title"
        >
          <span
            class="title"
            style="color: green;"
          >
            China
          </span>
           
          <span
            class="subtitle"
          >
            <div>
              Came here for a conference in 2005.
            </div>
          </span>
        </div>
         
        <div
          class="post-image"
        >
          <img
            alt="China"
            height="56px"
            loading="lazy"
            src="https://ik.imagekit.io/fireman35/pics/china_Gx-ZxMNsej.jpg?tr=w-112"
            width="56px"
          />
        </div>
      </div>
       
      <div
        class="divider"
      />
    </div>
    <div
      class="flex-row post-list__container"
    >
      <div
        class="flex-row justify-space-between mini-post"
      >
        <div
          class="flex-column post-title"
        >
          <span
            class="title"
            style="color: blue;"
          >
            Malaysia
          </span>
           
          <span
            class="subtitle"
          >
            <div>
              Came here to hike up a mountain in 2010.
            </div>
          </span>
        </div>
         
        <div
          class="post-image"
        >
          <img
            alt="Malaysia"
            height="56px"
            loading="lazy"
            src="https://ik.imagekit.io/fireman35/pics/malaysia_ZeAVPhILkr87.jpg?tr=w-112"
            width="56px"
          />
        </div>
      </div>
       
      <div
        class="divider"
      />
    </div>
    <div
      class="flex-row post-list__container"
    >
      <div
        class="flex-row justify-space-between mini-post"
      >
        <div
          class="flex-column post-title"
        >
          <span
            class="title"
            style="color: green;"
          >
            Indonesia
          </span>
           
          <span
            class="subtitle"
          >
            <div>
              Came here for a conference in 2005.
            </div>
          </span>
        </div>
         
        <div
          class="post-image"
        >
          <img
            alt="Indonesia"
            height="56px"
            loading="lazy"
            src="https://ik.imagekit.io/fireman35/pics/indonesia_vvghKPrEN.jpg?tr=w-112"
            width="56px"
          />
        </div>
      </div>
       
      <div
        class="divider"
      />
    </div>
    <div
      class="flex-row post-list__container"
    >
      <div
        class="flex-row justify-space-between mini-post"
      >
        <div
          class="flex-column post-title"
        >
          <span
            class="title"
            style="color: blue;"
          >
            Nepal
          </span>
           
          <span
            class="subtitle"
          >
            <div>
              Came here for a school trip in 1999.
            </div>
          </span>
        </div>
         
        <div
          class="post-image"
        >
          <img
            alt="Nepal"
            height="56px"
            loading="lazy"
            src="https://ik.imagekit.io/fireman35/pics/nepal_Ef8ASdGYr.jpg?tr=w-112"
            width="56px"
          />
        </div>
      </div>
       
      <div
        class="divider"
      />
    </div>
    <div
      class="flex-row post-list__container"
    >
      <div
        class="flex-row justify-space-between mini-post"
      >
        <div
          class="flex-column post-title"
        >
          <span
            class="title"
            style="color: blue;"
          >
            Japan
          </span>
           
          <span
            class="subtitle"
          >
            <div>
              Came here to learn pottery in 2011.
            </div>
          </span>
        </div>
         
        <div
          class="post-image"
        >
          <img
            alt="Japan"
            height="56px"
            loading="lazy"
            src="https://ik.imagekit.io/fireman35/pics/japan_0nwm163zE.jpg?tr=w-112"
            width="56px"
          />
        </div>
      </div>
       
      <!---->
    </div>
  </div>
</div>
`;
